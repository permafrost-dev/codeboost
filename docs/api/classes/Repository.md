[codeboost](../README.md) / [Exports](../modules.md) / Repository

# Class: Repository

## Table of contents

### Constructors

-   [constructor](Repository.md#constructor)

### Properties

-   [name](Repository.md#name)
-   [owner](Repository.md#owner)
-   [path](Repository.md#path)

### Accessors

-   [git](Repository.md#git)
-   [info](Repository.md#info)

### Methods

-   [checkout](Repository.md#checkout)
-   [clone](Repository.md#clone)
-   [createFork](Repository.md#createfork)
-   [currentBranch](Repository.md#currentbranch)
-   [defaultBranch](Repository.md#defaultbranch)
-   [fullRepositoryName](Repository.md#fullrepositoryname)
-   [localBranches](Repository.md#localbranches)
-   [onBranch](Repository.md#onbranch)
-   [prepare](Repository.md#prepare)
-   [pushToFork](Repository.md#pushtofork)

## Constructors

### constructor

• **new Repository**(`fullRepositoryName`, `repositoryStoragePath`)

#### Parameters

| Name                    | Type     |
| :---------------------- | :------- |
| `fullRepositoryName`    | `string` |
| `repositoryStoragePath` | `string` |

#### Defined in

[lib/Repository.ts:34](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L34)

## Properties

### name

• **name**: `string`

#### Defined in

[lib/Repository.ts:10](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L10)

---

### owner

• **owner**: `string`

#### Defined in

[lib/Repository.ts:11](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L11)

---

### path

• **path**: `string`

#### Defined in

[lib/Repository.ts:12](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L12)

## Accessors

### git

• `get` **git**(): `SimpleGit`

#### Returns

`SimpleGit`

#### Defined in

[lib/Repository.ts:22](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L22)

---

### info

• `get` **info**(): [`RepositoryInfo`](../interfaces/RepositoryInfo.md)

#### Returns

[`RepositoryInfo`](../interfaces/RepositoryInfo.md)

#### Defined in

[lib/Repository.ts:15](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L15)

## Methods

### checkout

▸ **checkout**(`branchName`): `Promise`<`void`\>

#### Parameters

| Name         | Type     |
| :----------- | :------- |
| `branchName` | `string` |

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:81](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L81)

---

### clone

▸ **clone**(): `Promise`<`void`\>

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:43](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L43)

---

### createFork

▸ **createFork**(): `Promise`<`void`\>

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:101](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L101)

---

### currentBranch

▸ **currentBranch**(): `Promise`<`string`\>

#### Returns

`Promise`<`string`\>

#### Defined in

[lib/Repository.ts:73](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L73)

---

### defaultBranch

▸ **defaultBranch**(): `Promise`<`string`\>

#### Returns

`Promise`<`string`\>

#### Defined in

[lib/Repository.ts:96](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L96)

---

### fullRepositoryName

▸ **fullRepositoryName**(): `string`

#### Returns

`string`

#### Defined in

[lib/Repository.ts:124](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L124)

---

### localBranches

▸ **localBranches**(): `Promise`<`BranchSummary`\>

#### Returns

`Promise`<`BranchSummary`\>

#### Defined in

[lib/Repository.ts:67](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L67)

---

### onBranch

▸ **onBranch**(`branchName`): `Promise`<`boolean`\>

#### Parameters

| Name         | Type     |
| :----------- | :------- |
| `branchName` | `string` |

#### Returns

`Promise`<`boolean`\>

#### Defined in

[lib/Repository.ts:77](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L77)

---

### prepare

▸ **prepare**(): `Promise`<`void`\>

Checks out the default branch and pulls down the latest changes

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:58](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L58)

---

### pushToFork

▸ **pushToFork**(`branchName`): `Promise`<`void`\>

#### Parameters

| Name         | Type     |
| :----------- | :------- |
| `branchName` | `string` |

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:120](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L120)
