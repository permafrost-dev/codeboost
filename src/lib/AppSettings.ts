/* It initializes a git instance */
export interface AppSettings {
    github_token: string;
    repository_storage_path: string;
    use_forks: boolean;
    use_pull_requests: boolean;
}

export function loadSettings(filename: string): AppSettings {
    const settings = require(filename).default;

    return transformSettings(settings);
}

export function transformSettings(settings: AppSettings): AppSettings {
    // load settings from environment variable references starting with '$'
    for (const key in settings) {
        if (typeof settings[key] === 'string' && settings[key].startsWith('$')) {
            if (typeof process.env[settings[key].slice(1)] !== 'undefined') {
                settings[key] = process.env[settings[key].slice(1)];
            }
        }
    }

    return settings;
}
