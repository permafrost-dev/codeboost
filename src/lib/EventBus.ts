import EventEmitter2 from 'eventemitter2';

export default class EventBus {
    public emitter: EventEmitter2;

    constructor() {
        this.emitter = new EventEmitter2();
    }

    public on(event: string, listener: (...args: any[]) => void) {
        return this.emitter.on(event, listener);
    }

    public off(event: string, listener: (...args: any[]) => void) {
        return this.emitter.off(event, listener);
    }

    public emit(event: string, ...args: any[]) {
        return this.emitter.emit(event, ...args);
    }

    public once(event: string, listener: (...args: any[]) => void) {
        return this.emitter.once(event, listener);
    }
}

export const eventbus = new EventBus();
