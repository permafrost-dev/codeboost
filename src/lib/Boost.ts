import { Repository } from '@/lib/Repository';
import { BoostConfiguration } from '@/types/BoostConfiguration';
import { existsSync } from 'fs';
import { SimpleGit } from 'simple-git';

export interface BoostScriptHandlerParameters {
    boost: Boost;
    repository: Repository;
    git: SimpleGit;
}

export type BoostScriptHandler = (params: BoostScriptHandlerParameters) => any;

export class Boost {
    protected config!: BoostConfiguration;

    public path!: string;
    public id!: string;
    public version!: string;
    public repositoryLimits!: {
        maxRunsPerVersion: number;
        minutesBetweenRuns: number;
    };
    public pullRequest!: {
        title: string;
        body: string;
        branch: string;
    };
    public scripts!: any[];
    public actions!: any[];

    constructor(path: string, config: BoostConfiguration) {
        this.path = `${path}/${config.version}`;
        this.id = config.id;
        this.version = config.version;
        this.repositoryLimits = {
            maxRunsPerVersion: config.repository_limits.max_runs_per_version,
            minutesBetweenRuns: config.repository_limits.minutes_between_runs,
        };
        this.pullRequest = config.pull_request;
        this.scripts = this.loadScripts(config.scripts.files);
        this.actions = config.actions;

        this.config = Object.assign({}, config);
    }

    public loadScripts(scripts: string[]) {
        const result: any[] = [];

        for (const script of scripts) {
            const fn = `${this.path}/${script}`;

            if (!existsSync(fn)) {
                throw new Error(`Boost script not found: ${fn}`);
            }

            result.push(require(fn).handler);
        }

        return result;
    }

    public async run(repository: Repository) {
        const params: BoostScriptHandlerParameters = {
            boost: this,
            repository,
            git: repository.git,
        };

        if (!this.config.scripts.parallel) {
            for (const script of this.scripts) {
                await script(params);
            }
            return;
        }

        await Promise.all(this.scripts.map(script => script(params)));
    }
}
