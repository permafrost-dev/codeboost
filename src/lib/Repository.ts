import { parseFullRepositoryName } from '@/lib/stringHelpers';
import { execSync } from 'child_process';
import { existsSync, mkdirSync } from 'fs';
import { dirname } from 'path';
import simpleGit, { SimpleGit } from 'simple-git';

export class Repository {
    public name: string;
    public owner: string;
    public path: string;
    protected gitInstance!: SimpleGit;

    public get git() {
        if (this.gitInstance === undefined) {
            this.gitInstance = simpleGit({
                baseDir: this.path,
                binary: 'git',
                maxConcurrentProcesses: 2,
                trimmed: false,
            });
        }
        return this.gitInstance;
    }

    constructor(fullRepositoryName: string, repositoryStoragePath: string) {
        const { owner, name } = parseFullRepositoryName(fullRepositoryName);

        this.name = name;
        this.owner = owner;

        this.path = `${repositoryStoragePath}/${fullRepositoryName}`;
    }

    public async clone() {
        const parentDir = dirname(this.path);

        if (!existsSync(parentDir)) {
            mkdirSync(parentDir, { recursive: true });
        }

        if (!existsSync(this.path)) {
            execSync(`git -C '${parentDir}' clone git@github.com:${this.owner}/${this.name}.git`, { stdio: 'pipe' });
        }
    }

    public async init() {
        //
    }
}
