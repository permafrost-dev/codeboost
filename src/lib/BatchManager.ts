import { Boost } from '@/lib/Boost';

export class BatchManager {
    public data: any[] = [];

    constructor(public filename: string) {
        this.data = require(filename);
    }

    public getBatch(boost: Boost, size: number) {
        //
    }

    protected getUsableDataset(boost: Boost) {
        return this.data.filter(item => boost.canRunOnRepository(item.repository));
    }

    protected getItemState(item: any) {
        return item.state || {};
    }
}
