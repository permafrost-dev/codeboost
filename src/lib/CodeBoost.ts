import { Boost } from '@/lib/Boost';
import { createOctokit } from '@/lib/github';
import { Repository } from '@/lib/Repository';
import { BoostConfiguration } from '@/types/BoostConfiguration';
import { Octokit } from '@octokit/core';
import { existsSync } from 'fs';
import simpleGit, { SimpleGit, SimpleGitOptions } from 'simple-git';

export class CodeBoost {
    protected repository!: Repository;
    protected repositoryDir: string = '';
    protected gitInstance!: SimpleGit;
    public octokit!: Octokit;

    public get git() {
        if (!this.gitInstance) {
            this.initGit();
        }

        return this.gitInstance;
    }

    public get repositoryPath() {
        return this.repositoryDir;
    }

    public set repositoryPath(value: string) {
        this.repositoryDir = value;

        if (existsSync(value)) {
            this.initGit();
        }
    }

    /**
     * Initializes a git instance, should be called once the repository path is set
     * to a path that exists. Called automatically when setting the repository path.
     */
    public initGit() {
        const createSimpleGitOptions = (): Partial<SimpleGitOptions> => ({
            baseDir: this.repositoryPath,
            binary: 'git',
            maxConcurrentProcesses: 2,
            trimmed: false,
        });

        this.gitInstance = simpleGit(createSimpleGitOptions());
    }

    public async init(repository: Repository) {
        this.repository = repository;
        this.repositoryPath = repository.path;
        this.octokit = createOctokit();

        const bc = this.loadBoostConfiguration('test-one');
        const boost = new Boost(`${__dirname}/boosts/${bc.id}`, bc);

        console.log({ bc, boost });

        await boost.run(repository);
    }

    public loadBoostConfiguration(id: string) {
        const config = require(`${__dirname}/boosts/${id}/boost.js`).default;

        return <BoostConfiguration>config;
    }
}
