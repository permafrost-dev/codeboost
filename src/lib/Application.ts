import { AppSettings, loadSettings } from '@/lib/AppSettings';
import { CodeBoost } from '@/lib/CodeBoost';
import { initOctokit } from '@/lib/github';
import { Repository } from '@/lib/Repository';

export class Application {
    public settings!: AppSettings;
    public codeboost!: CodeBoost;
    public repositoryName!: string;

    async init() {
        this.settings = loadSettings(`${__dirname}/settings.js`);
        initOctokit(this.settings.github_token);
        const repo = new Repository(this.repositoryName, this.settings.repository_storage_path);

        await repo.clone();

        this.codeboost = new CodeBoost();
        await this.codeboost.init(repo);
    }

    async execute(repoName: string, boostName: string) {
        this.repositoryName = repoName;
        await this.init();

        console.log('execute', repoName, boostName);
        console.log('settings:', this.settings);
    }
}
