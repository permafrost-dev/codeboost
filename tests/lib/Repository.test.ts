import { Repository } from '@/lib/Repository';

it('sets properties correctly on create', async () => {
    const repository = new Repository('owner/name', 'path');

    expect(repository).toBeInstanceOf(Repository);

    expect(repository.name).toBe('name');
    expect(repository.owner).toBe('owner');
    expect(repository.path).toBe('path/owner/name');

    expect(repository).toMatchSnapshot();
});
