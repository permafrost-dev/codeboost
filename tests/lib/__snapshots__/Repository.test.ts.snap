// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets properties correctly on create 1`] = `
Repository {
  "gitInstance": undefined,
  "name": "name",
  "owner": "owner",
  "path": "path/owner/name",
}
`;
